File: app/src/main/java/com/tommihirvonen/exifnotes/adapters/FilmManufacturerAdapter.java
package com.tommihirvonen.exifnotes.adapters;

import android.content.Context;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.Transformation;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tommihirvonen.exifnotes.R;
import com.tommihirvonen.exifnotes.datastructures.FilmStock;
import com.tommihirvonen.exifnotes.utilities.FilmDbHelper;

import java.util.List;

public class FilmManufacturerAdapter extends RecyclerView.Adapter<FilmManufacturerAdapter.ViewHolder> {

    private Context context;
    private List<String> manufacturers;
    private final SparseBooleanArray expandedManufacturers = new SparseBooleanArray();
    private final SparseBooleanArray expandAnimations = new SparseBooleanArray();
    private int currentExpandedIndex = -1;
    private FilmDbHelper database;
    private OnFilmStockSelectedListener listener;

    public interface OnFilmStockSelectedListener {
        void onFilmStockSelected(FilmStock filmStock);
    }

    public FilmManufacturerAdapter(@NonNull final Context context,
                                   @NonNull final List<String> manufacturers,
                                   @NonNull final OnFilmStockSelectedListener listener) {
        this.context = context;
        this.manufacturers = manufacturers;
        this.listener = listener;
        database = FilmDbHelper.getInstance(context);
        setHasStableIds(true);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        final TextView manufacturerTextView;
        final View manufacturerLayout;
        final View expandLayout;
        final ImageView expandButton;
        final RecyclerView filmStocksRecyclerView;
        ViewHolder(final View itemView) {
            super(itemView);
            manufacturerTextView = itemView.findViewById(R.id.text_view_manufacturer_name);
            manufacturerLayout = itemView.findViewById(R.id.layout_manufacturer);
            expandLayout = itemView.findViewById(R.id.layout_expand);
            expandButton = itemView.findViewById(R.id.image_view_expand);
            filmStocksRecyclerView = itemView.findViewById(R.id.recycler_view_film_stocks);

        }
    }

    @NonNull
    @Override
    public FilmManufacturerAdapter.ViewHolder onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType) {
        final View view = LayoutInflater.from(context).inflate(R.layout.item_film_manufacturer, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull final ViewHolder holder, final int position) {
        final String manufacturer = manufacturers.get(position);

        final List<FilmStock> filmStocks = database.getAllFilmStocks(manufacturer);
        final FilmStockAdapter adapter = new FilmStockAdapter(filmStocks);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        holder.filmStocksRecyclerView.setLayoutManager(layoutManager);
        holder.filmStocksRecyclerView.setAdapter(adapter);
        holder.manufacturerTextView.setText(manufacturer);
        holder.manufacturerLayout.setOnClickListener(v -> {
            toggleManufacturer(position);
            expandOrCollapseManufacturer(holder, position);
        });
        expandOrCollapseManufacturer(holder, position);
    }

    @Override
    public int getItemCount() {
        return manufacturers.size();
    }

    @Override
    public long getItemId(int position) {
        return manufacturers.get(position).hashCode();
    }

    private void toggleManufacturer(final int position) {
        currentExpandedIndex = position;
        if (expandedManufacturers.get(position, false)) {
            expandedManufacturers.delete(position);
            expandAnimations.delete(position);
        } else {
            expandedManufacturers.put(position, true);
            expandAnimations.put(position, true);
        }
    }

    private void expandOrCollapseManufacturer(final ViewHolder holder, final int position) {
        final boolean animate = currentExpandedIndex == position;
        if (expandedManufacturers.get(position, false)) {
            toggleArrow(holder.expandButton, true, animate);
            toggleLayout(holder.expandLayout, true, animate);
        } else {
            toggleArrow(holder.expandButton, false, animate);
            toggleLayout(holder.expandLayout, false, animate);
        }
        if (animate) currentExpandedIndex = -1;
    }

    private static void toggleLayout(final View view, final boolean isExpanded,
                                     final boolean animate) {
        if (isExpanded) {
            expand(view, animate);
        } else {
            collapse(view, animate);
        }
    }

    private static void toggleArrow(final View view, final boolean isExpanded, final boolean animate) {
        if (isExpanded && animate) {
            view.animate().setDuration(200).rotation(180);
        } else if (isExpanded) {
            view.setRotation(180);
        } else if (animate) {
            view.animate().setDuration(200).rotation(0);
        } else {
            view.setRotation(0);
        }
    }

    private static void expand(final View view, final boolean animate) {
        if (animate) {
            // Call view.measure() before calling view.getMeasuredHeight()
            view.measure(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT);
            final int actualHeight = view.getMeasuredHeight();
            final int currentHeight = view.getLayoutParams().height;
            view.setVisibility(View.VISIBLE);
            final Animation animation = new Animation() {
                @Override
                protected void applyTransformation(float interpolatedTime, Transformation t) {
                    // interpolatedTime == 1 => the animation has reached its end
                    if (interpolatedTime == 1) {
                        view.getLayoutParams().height = ViewGroup.LayoutParams.WRAP_CONTENT;
                    } else {
                        view.getLayoutParams().height = (int) (currentHeight + (actualHeight - currentHeight) * interpolatedTime);
                    }
                    view.requestLayout();
                }
            };
            animation.setDuration((long) (actualHeight / view.getContext().getResources().getDisplayMetrics().density));
            view.startAnimation(animation);
        } else {
            view.setVisibility(View.VISIBLE);
        }
    }

    private static void collapse(final View view, final boolean animate) {
        if (animate) {
            final int actualHeight = view.getMeasuredHeight();
            final Animation animation = new Animation() {
                @Override
                protected void applyTransformation(float interpolatedTime, Transformation t) {
                    // interpolatedTime == 1 => the animation has reached its end
                    if (interpolatedTime == 1) {
                        view.setVisibility(View.GONE);
                    } else {
                        view.getLayoutParams().height = actualHeight - (int) (actualHeight * interpolatedTime);
                        view.requestLayout();
                    }
                }
            };
            animation.setDuration((long) (actualHeight / view.getContext().getResources().getDisplayMetrics().density));
            view.startAnimation(animation);
        } else {
            view.setVisibility(View.GONE);
            view.getLayoutParams().height = 0;
        }
    }

    private class FilmStockAdapter extends RecyclerView.Adapter<FilmStockAdapter.ViewHolder> {
        private List<FilmStock> filmStocks;
        FilmStockAdapter(@NonNull final List<FilmStock> filmStocks) {
            this.filmStocks = filmStocks;
            setHasStableIds(true);
        }
        class ViewHolder extends RecyclerView.ViewHolder {
            final TextView filmStockTextView;
            final LinearLayout filmStockLayout;
            ViewHolder(final View itemView) {
                super(itemView);
                filmStockTextView = itemView.findViewById(R.id.text_view_film_stock);
                filmStockLayout = itemView.findViewById(R.id.layout_film_stock);
            }
        }
        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            final View view = LayoutInflater.from(context).inflate(R.layout.item_film_stock, parent, false);
            return new FilmStockAdapter.ViewHolder(view);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            final FilmStock filmStock = filmStocks.get(position);
            holder.filmStockTextView.setText(filmStock.getModel());
            holder.filmStockLayout.setOnClickListener(v -> listener.onFilmStockSelected(filmStock));
        }

        @Override
        public int getItemCount() {
            return filmStocks.size();
        }

        @Override
        public long getItemId(int position) {
            return filmStocks.get(position).getId();
        }
    }

}

File: app/src/main/java/com/tommihirvonen/exifnotes/adapters/FilmManufacturerAdapter.kt
/*
 * Exif Notes
 * Copyright (C) 2022  Tommi Hirvonen
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.tommihirvonen.exifnotes.adapters

import android.content.Context
import android.util.SparseBooleanArray
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.Transformation
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tommihirvonen.exifnotes.databinding.ItemFilmManufacturerBinding
import com.tommihirvonen.exifnotes.databinding.ItemFilmStockBinding
import com.tommihirvonen.exifnotes.datastructures.FilmStock
import java.util.*

class FilmManufacturerAdapter(
        private val context: Context,
        private val listener: (FilmStock?) -> Unit?)
    : RecyclerView.Adapter<FilmManufacturerAdapter.ViewHolder>() {

    private val expandedManufacturers = SparseBooleanArray()
    private val expandAnimations = SparseBooleanArray()
    private var currentExpandedIndex = -1
    private var filmStocksMap: Map<String?, List<FilmStock>> = emptyMap()
    private var manufacturers: List<String?> = emptyList()

    init { setHasStableIds(true) }

    inner class ViewHolder(val binding: ItemFilmManufacturerBinding) : RecyclerView.ViewHolder(binding.root)

    fun setFilmStocks(filmStocks: List<FilmStock>) {
        filmStocksMap = filmStocks.groupBy { it.make }
        manufacturers = filmStocksMap.map { it.key }.sortedBy { it?.lowercase(Locale.ROOT) }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val inflater = LayoutInflater.from(context)
        return ViewHolder(ItemFilmManufacturerBinding.inflate(inflater, parent, false))
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val manufacturer = manufacturers[position]
        val filmStocks = filmStocksMap[manufacturer] ?: return
        val adapter = FilmStockAdapter(filmStocks)
        val layoutManager = LinearLayoutManager(context)
        holder.binding.recyclerViewFilmStocks.layoutManager = layoutManager
        holder.binding.recyclerViewFilmStocks.adapter = adapter
        holder.binding.textViewManufacturerName.text = manufacturer
        holder.binding.layoutManufacturer.setOnClickListener {
            toggleManufacturer(position)
            expandOrCollapseManufacturer(holder, position)
        }
        expandOrCollapseManufacturer(holder, position)
    }

    override fun getItemCount(): Int {
        return manufacturers.size
    }

    override fun getItemId(position: Int): Long {
        return manufacturers[position].hashCode().toLong()
    }

    private fun toggleManufacturer(position: Int) {
        currentExpandedIndex = position
        if (expandedManufacturers[position, false]) {
            expandedManufacturers.delete(position)
            expandAnimations.delete(position)
        } else {
            expandedManufacturers.put(position, true)
            expandAnimations.put(position, true)
        }
    }

    private fun expandOrCollapseManufacturer(holder: ViewHolder, position: Int) {
        val animate = currentExpandedIndex == position
        if (expandedManufacturers[position, false]) {
            toggleArrow(holder.binding.imageViewExpand, true, animate)
            toggleLayout(holder.binding.layoutExpand, true, animate)
        } else {
            toggleArrow(holder.binding.imageViewExpand, false, animate)
            toggleLayout(holder.binding.layoutExpand, false, animate)
        }
        if (animate) currentExpandedIndex = -1
    }

    private inner class FilmStockAdapter(private val filmStocks: List<FilmStock>)
        : RecyclerView.Adapter<FilmStockAdapter.ViewHolder>() {

        init { setHasStableIds(true) }

        inner class ViewHolder(val binding: ItemFilmStockBinding) : RecyclerView.ViewHolder(binding.root)

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val inflater = LayoutInflater.from(context)
            return ViewHolder(ItemFilmStockBinding.inflate(inflater, parent, false))
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val filmStock = filmStocks[position]
            holder.binding.textViewFilmStock.text = filmStock.model
            holder.binding.layoutFilmStock.setOnClickListener { listener(filmStock) }
        }

        override fun getItemCount(): Int {
            return filmStocks.size
        }

        override fun getItemId(position: Int): Long {
            return filmStocks[position].id
        }

    }

    companion object {
        private fun toggleLayout(view: View, isExpanded: Boolean,
                                 animate: Boolean) {
            if (isExpanded) {
                expand(view, animate)
            } else {
                collapse(view, animate)
            }
        }

        private fun toggleArrow(view: View, isExpanded: Boolean, animate: Boolean) {
            if (isExpanded && animate) {
                view.animate().setDuration(200).rotation(180f)
            } else if (isExpanded) {
                view.rotation = 180f
            } else if (animate) {
                view.animate().setDuration(200).rotation(0f)
            } else {
                view.rotation = 0f
            }
        }

        private fun expand(view: View, animate: Boolean) {
            // Call view.measure() before calling view.getMeasuredHeight()
            view.measure(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT)
            if (animate) {
                val actualHeight = view.measuredHeight
                val currentHeight = view.layoutParams.height
                view.visibility = View.VISIBLE
                val animation: Animation = object : Animation() {
                    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                        // interpolatedTime == 1 => the animation has reached its end
                        if (interpolatedTime == 1f) {
                            view.layoutParams.height = ViewGroup.LayoutParams.WRAP_CONTENT
                        } else {
                            view.layoutParams.height = (currentHeight + (actualHeight - currentHeight) * interpolatedTime).toInt()
                        }
                        view.requestLayout()
                    }
                }
                animation.duration = (actualHeight / view.context.resources.displayMetrics.density).toLong()
                view.startAnimation(animation)
            } else {
                view.layoutParams.height = view.measuredHeight
                view.visibility = View.VISIBLE
            }
        }

        private fun collapse(view: View, animate: Boolean) {
            if (animate) {
                val actualHeight = view.measuredHeight
                val animation: Animation = object : Animation() {
                    override fun applyTransformation(interpolatedTime: Float, t: Transformation) {
                        // interpolatedTime == 1 => the animation has reached its end
                        if (interpolatedTime == 1f) {
                            view.visibility = View.GONE
                        } else {
                            view.layoutParams.height = actualHeight - (actualHeight * interpolatedTime).toInt()
                            view.requestLayout()
                        }
                    }
                }
                animation.duration = (actualHeight / view.context.resources.displayMetrics.density).toLong()
                view.startAnimation(animation)
            } else {
                view.visibility = View.GONE
                view.layoutParams.height = 0
            }
        }
    }

}